capsules/extra/src/bmm150.rs

```diff
@@ -0,0 +1,230 @@
+                self.pending_data.set(true);
+                self.initialize()
+            }
+            _ => {
+                if !self.pending_data.get() {
+                    self.pending_data.set(true);
+                    self.start_measurement()
+                } else {
+                    Err(ErrorCode::BUSY)
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum State {
+    Suspend,
+    Sleep,
+    InitializeReading,
+    ReadMeasurement,
+    Read,
+}
+
+impl<'a, I: I2CDevice> I2CClient for BMM150<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        if let Err(i2c_err) = status {
+            self.state.set(State::Sleep);
+            self.buffer.replace(buffer);
+            self.ninedof_client
+                .map(|client| client.callback(i2c_err as usize, 0, 0));
+            return;
+        }
+
+        match self.state.get() {
+            State::InitializeReading => {
+                buffer[0] = Registers::DATAxLsb as u8;
+
+                if let Err((i2c_err, buffer)) = self.i2c.write(buffer, 1) {
+                    self.state.set(State::Sleep);
+                    self.buffer.replace(buffer);
+                    self.ninedof_client
+                        .map(|client| client.callback(i2c_err as usize, 0, 0));
+                } else {
+                    self.state.set(State::ReadMeasurement);
+                }
+            }
+            State::ReadMeasurement => {
+                if let Err((i2c_err, buffer)) = self.i2c.read(buffer, 8) {
+                    self.state.set(State::Sleep);
+                    self.buffer.replace(buffer);
+                    self.ninedof_client
+                        .map(|client| client.callback(i2c_err as usize, 0, 0));
+                } else {
+                    self.state.set(State::Read);
+                }
+            }
+            State::Read => {
+                let x_axis = ((buffer[1] as i16) << 5) | ((buffer[0] as i16) >> 3);
+                let y_axis = ((buffer[3] as i16) << 5) | ((buffer[2] as i16) >> 3);
+                let z_axis = ((buffer[5] as i16) << 7) | ((buffer[4] as i16) >> 1);
+
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                if self.pending_data.get() {
+                    self.pending_data.set(false);
+                    self.ninedof_client.map(|client| {
+                        client.callback(x_axis as usize, y_axis as usize, z_axis as usize)
+                    });
+                }
+
+                self.state.set(State::Sleep);
+            }
+            State::Sleep => {
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                if self.pending_data.get() {
+                    self.pending_data.set(false);
+                    self.ninedof_client.map(|client| client.callback(0, 0, 0));
+                }
+            }
```

Comment:
This particular state is only entered after the call to `initialize()`. Since there is no available data from the sensor yet, the 0 values act like dummy values.

Reply:
Initialize isn't in the HIL. So a client will get setup via the HIL, but then will get a callback that it doesn't expect if initialize is called.

We generally handle this by having initialize only called if necessary (say once a measurement is requested) or it just does not have a callback.